OPNET-282: Configure-ovs alternative implementation

cybertron opened the pull request:
This adds a new service which allows manual configuration of br-ex using NMState. The advantage here versus configure-ovs is that you are able to explicitly configure the bridge exactly as you want from day 1, and also use NMState-based tools to modify it on day 2.

The configurations are provided as a machine-config provided to the installer as part of the manifests. The machine-config needs to contain either a cluster-wide or machine-specific NMState configuration that will be applied on first boot at approximately the same time configure-ovs would normally run. Machine-specific configs are mapped to their nodes using the hostname.

Additionally, this includes a couple of other meaningful changes. First is a wait-for-primary-ip service that runs after nmstate.service to ensure the IP is usable before we start crio. Otherwise crio may fail to start initially, which then blocks kubelet. The other change of note is that the directory used to store these new configurations is added to the rebootless config list. This is because these configs are only applied on first boot and changes to them would have no effect anyway. More importantly, when scaling out a cluster it is necessary to add configs for the new nodes and we don't want to have to reboot the entire existing cluster when we do that.

**- What I did**

**- How to verify it**

**- Description for the changelog**
Added nmstate-configuration service that provides an alternative to ovs-configuration for creation of br-ex.

yuqi-zhang commented on pkg/daemon/update.go:
```diff
@@ -419,6 +419,12 @@ func calculatePostConfigChangeActionFromFileDiffs(diffFileSet []string) (actions
 		imageRegistryAuthFile,
 		"/var/lib/kubelet/config.json",
 	}
+	directoriesPostConfigChangeActionNone := []string{
```
Just as a note we're trying to move from a hard-coded list to an API here: https://github.com/openshift/api/pull/1764

So directories isn't part of our initially supported plans but we should eventually add support for this cc @djoshy

yuqi-zhang commented on templates/common/_base/files/nmstate-configuration.yaml:
```diff
@@ -0,0 +1,29 @@
+    fi
+
+    hostname=$(hostname -s)
+    hostname_file="/etc/nmstate/openshift/${hostname}.yml"
```
just curious, is it possible to add everything into one file and post process it later?

i.e. instead of having ip1.yaml ip2.yaml cluster.yaml, I have one big config.yaml, that then I specify within:
ip1.spec and ip2.spec

or something like that so it's a singular path, and the script itself processes when it does the copy/processing?

cybertron replied:
There's probably some way we could do it. I'm not sure we have yq available on the host so I don't know if we can use that, but we could probably split the file with some bash magic.

yuqi-zhang reviewed (commented):
Discussed this on slack, but to summarize it here:

What we're essentially proposing is a dual management method where the user is expected to provide a MachineConfig object (used by ignition) to do provisioning during install time, and for new nodes. Day 2 configuration is done via NodeNetworkConfigurationPolicy objects which are not actually part of the core payload.

This is not ideal as it introduces some issues and antipatterns:
1. some machineconfigs don't actually do anything if you update them, which is confusing for a user (I guess it technically does affect new nodes, but not existing ondes)
2. to update a network configuration I'd have to edit both the API object and the machineconfig separately, which is not ideal
3. we're doing oneshot configuration via bash via a template and a stamp file, which isn't really what machineconfigs should be used for (although I guess it technically is supported to do so)

So this can probably work but I think longer term we should plan on either adding in proper API + parsing or add a new CRD into openshift that can do the higher level management directly (so it translates to both machineconfigs and NodeNetworkConfigurationPolicy, for example)

cybertron commented on pkg/daemon/update.go:
```diff
@@ -419,6 +419,12 @@ func calculatePostConfigChangeActionFromFileDiffs(diffFileSet []string) (actions
+		// We probably don't want to use this exact path because NMState is
+		// planning to add a service that applies configs from it too, and we
+		// want to make sure our service is the only one processing the configs.
+		"/etc/nmstate",
```
I should change this to be /etc/nmstate/openshift. We don't necessarily want to ignore _everything_ in /etc/nmstate.

cybertron commented:
/test e2e-metal-ipi
/test e2e-metal-ipi-ovn-ipv6

jcaamano commented on templates/common/_base/files/nmstate-configuration.yaml:
```diff
@@ -5,25 +5,38 @@ contents:
     #!/bin/bash
     set -eux
 
+    # Clean up old config on behalf of mtu-migration
+    if ! systemctl -q is-enabled mtu-migration; then
+      echo "Cleaning up left over mtu migration configuration"
+      rm -rf /etc/cno/mtu-migration
+    fi
```
I guess this makes the check in configure-ovs dead code?

cybertron replied:
I suppose it does if we leave this service enabled everywhere.

jcaamano commented on templates/common/_base/files/wait-for-primary-ip.yaml:
```diff
@@ -0,0 +1,31 @@
+      exit 0
+    fi
+
+    # This logic is borrowed from configure-ovs.sh
```
I guess this makes the check in configure-ovs dead code?

cybertron replied:
Not in this case. We only run this if the new mechanism is used, so the configure-ovs version will still apply when we don't.

jcaamano replied:
Ah ok. Why do you need to use `/run/nodeip-configuration/primary-ip` specifically? If configure-ovs is doing it and it is platform agnostic, then we could basically use whatever it is using. I guess not problem if we limit this to baremetal though.

cybertron replied:
I think this couldn't work exactly the same way configure-ovs does because it's not all one service. In this case, we have nmstate-configuration.service providing configuration files for nmstate.service, then run wait-for-primary-ip.service to ensure we have a usable address.

The platform-agnostic solution is probably to split this into two services, one that runs before nmstate.service and stores off the primary IP, then this one that runs after to wait for it. I'll see how hard that would be to do now that we have a little more time to get this done.

jcaamano commented on pkg/daemon/update.go:
```diff
@@ -419,6 +419,12 @@ func calculatePostConfigChangeActionFromFileDiffs(diffFileSet []string) (actions
+		// We probably don't want to use this exact path because NMState is
+		// planning to add a service that applies configs from it too, and we
+		// want to make sure our service is the only one processing the configs.
+		"/etc/nmstate/openshift",
```
Why chose this dir vs, say, a directory out of a well known path? Potential for conflicts are low but I guess it would be less if we chose a different path.

cybertron replied:
Do you have a suggestion on another location? I don't particularly care where the files are, but since this is OpenShift-specific NMState configuration I thought an openshift directory under /etc/nmstate made sense.

cybertron replied:
I will note that we had a discussion about naming this something like "/etc/nmstate/techpreview-openshift" to make sure it's clear to anyone using this functionality that it is currently tech preview. We'll be documenting that too, but not everyone reads the docs, unfortunately.

jcaamano replied:
We already use paths like `/etc/cno` and `/etc/ovnk`. `/etc/mco` also exists and I guess a potentially new `/etc/openshift` comes to mind as well. If none of these seem plausible to you, then well I guess I don't have really any suggestion so we can go with what you have here.

jcaamano reviewed (commented):
In general, the networking aspect of this LGTM.

cybertron commented:
I just squashed some obsolete commits to make this easier to review. No functional changes yet.

jcaamano commented on templates/common/_base/units/nmstate-configuration.service.yaml:
```diff
@@ -0,0 +1,25 @@
+name: nmstate-configuration.service
+enabled: true
```
I mean to say platform type. As per the enhancement
```
Initially, this feature will target just baremetal clusters. Once we are
confident that the design is working as intended in what are usually the most
complex networking environments for OpenShift, we will extend support to all
platforms.
```

cybertron replied:
Okay, done.

qinqon reviewed (changes requested):
nits

jcaamano commented:
LGTM

qinqon commented:
/lgtm

cgoncalves commented:
LGTM.

cybertron commented on pkg/daemon/update.go:
```diff
@@ -429,7 +435,13 @@ func calculatePostConfigChangeActionFromFileDiffs(diffFileSet []string) (actions
 	}
 
 	actions = []string{postConfigChangeActionNone}
+path:
```
Yeah, it turns out this was needlessly complicated. I'm now reusing the same method that was already in place here.

cybertron commented on templates/common/_base/files/nmstate-configuration.yaml:
```diff
@@ -0,0 +1,29 @@
+    if [ -z $config_file ]; then
+      # No config to apply
+      exit 0
+    fi
```
Ha, so it is. Done.

cybertron commented on templates/common/_base/files/nmstate-configuration.yaml:
```diff
@@ -0,0 +1,29 @@
+    fi
+
+    hostname=$(hostname -s)
+    hostname_file="/etc/nmstate/openshift/${hostname}.yml"
```
Done.

cybertron commented on templates/common/_base/files/wait-for-primary-ip.yaml:
```diff
@@ -0,0 +1,31 @@
+    #!/bin/bash
+    set -eux
+
+    if [ ! -e /etc/nmstate/openshift/applied ]; then
+      # No need to do this if no NMState configuration was applied
+      exit 0
+    fi
```
This isn't for second boot, it's to avoid running this when the new mechanism is not in use at all.

cybertron commented on templates/common/_base/files/nmstate-configuration.yaml:
```diff
@@ -0,0 +1,29 @@
+    set -eux
+
+    if [ -e /etc/nmstate/openshift/applied ]; then
+      # Already done, don't apply again
```
Sure, done.

qiowang721 commented on templates/common/_base/files/nmstate-configuration.yaml:
```diff
@@ -0,0 +1,42 @@
+    if [ -s "$src_path/$hostname_file" ]; then
+      config_file=$hostname_file
+    fi
+    if [ -s "$src_path/$cluster_file" ]; then
```
hi @cybertron if i understand correctly, hostname_file is node-specific configuration file and cluster_file is cluster-wide configuration file, right? if so, seems the cluster-wide one will take precedence, does it conflict with the description in enhancement?

cybertron replied:
You're right. This should be fixed by the change in logic in the next version of this.

qiowang721 commented:
Hi @cybertron I tried to build cluster with the PR via cluster-bot today, but got 'pull request openshift/machine-config-operator#4249 needs to be rebased to branch master', could you please rebase it? Thanks a lot.

cybertron commented:
Rebased and addressed most of the comments. I've added a couple of topics to the next sync meeting, but I think we move forward with this as-is and address the outstanding questions later.

qinqon commented:
/lgtm

yuqi-zhang reviewed (approved):
Approving from the MCO perspective

qiowang721 commented:
/hold

yuqi-zhang commented:
Ah, sorry, I may have jumped the gun on the approval. @qiowang721 for any issues, maybe we can follow up as bugs?

qiowang721 commented:
> Ah, sorry, I may have jumped the gun on the approval. @qiowang721 for any issues, maybe we can follow up as bugs?

np, i opened a bug for it~
